// expo-cli/build/commands/generate-universal.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _toConsumableArray2;

function _load_toConsumableArray() {
  return _toConsumableArray2 = _interopRequireDefault(require('babel-runtime/helpers/toConsumableArray'));
}

var _regenerator;

function _load_regenerator() {
  return _regenerator = _interopRequireDefault(require('babel-runtime/regenerator'));
}

var _promise;

function _load_promise() {
  return _promise = _interopRequireDefault(require('babel-runtime/core-js/promise'));
}

var _asyncToGenerator2;

function _load_asyncToGenerator() {
  return _asyncToGenerator2 = _interopRequireDefault(require('babel-runtime/helpers/asyncToGenerator'));
}

var _path = _interopRequireDefault(require('path'));

var _child_process = _interopRequireDefault(require('child_process'));

var _targz;

function _load_targz() {
  return _targz = _interopRequireDefault(require('targz'));
}

var _fs = _interopRequireDefault(require('fs'));

var _fsExtra;

function _load_fsExtra() {
  return _fsExtra = _interopRequireDefault(require('fs-extra'));
}

var _replace;

function _load_replace() {
  return _replace = _interopRequireDefault(require('replace'));
}

var _CommandError;

function _load_CommandError() {
  return _CommandError = _interopRequireDefault(require('../CommandError'));
}

var _prompt;

function _load_prompt() {
  return _prompt = _interopRequireDefault(require('../prompt'));
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var NPM_TEMPLATE_VERSION = '^1.0.1';
var TEMP_DIR_NAME = 'temp-expo-module-template';
var archive = void 0;

var decompress = function () {
  var _ref = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee() {
    return (_regenerator || _load_regenerator()).default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            return _context.abrupt('return', new (_promise || _load_promise()).default(function (resolve, reject) {
              (_targz || _load_targz()).default.decompress({
                src: archive,
                dest: TEMP_DIR_NAME
              }, function (error) {
                if (error) {
                  reject(error);
                } else {
                  resolve(null);
                }
              });
            }));

          case 1:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, undefined);
  }));

  return function decompress() {
    return _ref.apply(this, arguments);
  };
}();

exports.default = function (program) {
  program.command('generate <template>').description('Generate a module from a template.').asyncAction(function () {
    var _ref2 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee2(template) {
      var configuration, javaDir, placeholderPath;
      return (_regenerator || _load_regenerator()).default.wrap(function _callee2$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              if (!(template !== 'universal')) {
                _context2.next = 2;
                break;
              }

              throw new (_CommandError || _load_CommandError()).default('UNKNOWN_TEMPLATE', 'Template not found: \'' + template + '\'. Valid options are: universal');

            case 2:
              _context2.next = 4;
              return (0, (_prompt || _load_prompt()).default)([{
                name: 'jsName',
                message: 'How would you like to call your module in JS/npm? (eg. expo-camera)'
              }, {
                name: 'podName',
                message: 'How would you like to call your module in CocoaPods? (eg. EXCamera) (leave empty to not include iOS part)'
              }, {
                name: 'javaModule',
                message: 'How would you like to call your module in Java? (eg. expo.modules.camera)'
              }]);

            case 4:
              configuration = _context2.sent;


              archive = _child_process.default.execSync('npm pack expo-module-template@' + NPM_TEMPLATE_VERSION).toString().slice(0, -1);
              if (!_fs.default.existsSync(TEMP_DIR_NAME)) {
                _fs.default.mkdirSync(TEMP_DIR_NAME);
              }
              _context2.next = 9;
              return decompress();

            case 9:

              _fs.default.unlinkSync(archive);
              _context2.next = 12;
              return (_fsExtra || _load_fsExtra()).default.copy(_path.default.join(TEMP_DIR_NAME, 'package'), '' + configuration.jsName);

            case 12:
              _context2.next = 14;
              return (_fsExtra || _load_fsExtra()).default.remove(TEMP_DIR_NAME);

            case 14:
              if (!configuration.podName) {
                _context2.next = 22;
                break;
              }

              _fs.default.renameSync(_path.default.join(configuration.jsName, 'ios', 'EXModuleTemplate.podspec'), _path.default.join(configuration.jsName, 'ios', configuration.podName + '.podspec'));

              _fs.default.renameSync(_path.default.join(configuration.jsName, 'ios', 'EXModuleTemplate', 'EXModuleTemplate.h'), _path.default.join(configuration.jsName, 'ios', 'EXModuleTemplate', configuration.podName + '.h'));

              _fs.default.renameSync(_path.default.join(configuration.jsName, 'ios', 'EXModuleTemplate', 'EXModuleTemplate.m'), _path.default.join(configuration.jsName, 'ios', 'EXModuleTemplate', configuration.podName + '.m'));

              _fs.default.renameSync(_path.default.join(configuration.jsName, 'ios', 'EXModuleTemplate'), _path.default.join(configuration.jsName, 'ios', '' + configuration.podName));

              _fs.default.renameSync(_path.default.join(configuration.jsName, 'ios', 'EXModuleTemplate.xcodeproj'), _path.default.join(configuration.jsName, 'ios', configuration.podName + '.xcodeproj'));
              _context2.next = 24;
              break;

            case 22:
              _context2.next = 24;
              return (_fsExtra || _load_fsExtra()).default.remove(_path.default.join(configuration.jsName, 'ios'));

            case 24:

              (0, (_replace || _load_replace()).default)({
                regex: 'expo-module-template',
                replacement: configuration.jsName,
                paths: [configuration.jsName],
                recursive: true,
                silent: true
              });

              (0, (_replace || _load_replace()).default)({
                regex: 'expo.modules.template',
                replacement: configuration.javaModule,
                paths: [configuration.jsName],
                recursive: true,
                silent: true
              });

              (0, (_replace || _load_replace()).default)({
                regex: 'EXModuleTemplate',
                replacement: configuration.podName,
                paths: [configuration.jsName],
                recursive: true,
                silent: true
              });

              (0, (_replace || _load_replace()).default)({
                regex: '"version": ".*",',
                replacement: '"version": "1.0.0",',
                paths: [_path.default.join(configuration.jsName, 'package.json')],
                recursive: false,
                silent: true
              });

              javaDir = _path.default.join.apply(_path.default, [configuration.jsName, 'android', 'src', 'main', 'java'].concat((0, (_toConsumableArray2 || _load_toConsumableArray()).default)(configuration.javaModule.split('.'))));

              (_fsExtra || _load_fsExtra()).default.mkdirpSync(javaDir);

              placeholderPath = _path.default.join(javaDir, 'Placeholder.java');

              _fs.default.appendFileSync(placeholderPath, 'package ' + configuration.javaModule + ';\n');
              _fs.default.appendFileSync(placeholderPath, 'class Placeholder {}');

            case 33:
            case 'end':
              return _context2.stop();
          }
        }
      }, _callee2, undefined);
    }));

    return function (_x) {
      return _ref2.apply(this, arguments);
    };
  }());
};

module.exports = exports['default'];
